### ISSUE TITLE
Recover after error

### ISSUE THREAD

**Issue by Felixrccs** [libAtoms/workflow#257]
If I want to calculate large amounts of DFT calculations over wfl on the cluster I ran into the problem, that I reached the wall-time and the job failed. In that case I have to go through the calculation directories and recover the calculations manually.

Would it be possible to add a feature where I can give wfl an internal wall-time and it stops after i.e. 20 h returns me one xyz with successful and another xyz with failed/ non-finished calculations?

**Comment by bernstei** [comment]
With some limitations, it may actually be more straightforward than I first thought. The actual timeout mechanism for the calculation is, I think (need to test the details), pretty straightforward with `multiprocessing.pool.imap`, which `autoparallelize` uses.  My current challenge is that there's currently no access to the _input_ configurations when looping over the _outputs_ in `autoparallelize`, so no easy way to return an annotated input config when the timeout happens.

**Comment by Felixrccs** [comment]
> Maybe you should describe your use case more fully (number of configs, sizes of cells, heterogeneity of sizes, range of times per calculation, etc) to see if there's a way of dealing with it without adding this complexity.

So I have to recalculate my trainingset with tighter DFT parameters (about 1000 configs). I'm looking at metal-oxid surfaces that have between 80 and 200 atoms. The DFT single point take normally between 1 and 3 hours per single point on one node. I calculate 10 subproccesses on ten nodes, 50 structures take about 15 hours of wall time in total. However from time to time there are a few amorphous structures that need up 10 hours for a single point. Our cluster has a 24 hour submission limit and in one case every DFT was finished except for one complex config.

I solved it for the my case by parallelizing the single points within VASP over multiple nodes.

But I think this would probably still be a good feature regarding user-friendliness.

>  My current challenge is that there's currently no access to the input configurations when looping over the outputs in autoparallelize, so no easy way to return an annotated input config when the timeout happens.

I will think about this myself. If you have something to test, I can try it out.

**Comment by bernstei** [comment]
So by far the cleanest solution to this is to use the queuing system for what it's designed for, i.e. to run multiple jobs side by side or in sequence as resources become available, and the easiest way to do that is one queued job per configuration.

Is that an option (perhaps we can make the time and/or number of nodes settable on a per-config basis), or do you have a problem such as a maliciously-configured queuing system with a small limit on the number of submitted jobs (some UK HPC - archer, I think) ?

**Comment by Felixrccs** [comment]
No I can split it up in one job per DFT (mpg HPC queing systeme is performing fine).

**Comment by bernstei** [comment]
> I think this may actually be pretty straightforward. I'll test my idea it once I finish testing the kspacing thing.

OK - on further thought, depending on what kind of timeouts you want to deal with, this could be quite messy. I think this needs a deep discussion first, and going through the possible use cases very carefully.  I'm happy to do it, but I won't delay my other PR (the DFT calculator kspacing and non-periodic cell update) for it.

Also, I have a question:

>  I calculate 10 subproccesses on ten nodes

What _exactly_ do you mean by this statement?  Are you trying to do pool-based parallelism with 10 subprocesses in a job that uses 10 nodes?  Can you describe more precisely how you are setting up the wfl script that's doing the database re-evaluation to try to achieve this?  I ask because the normal `multiprocessing.pool` based autoparallelism doesn't work on multiple nodes.

**Comment by Felixrccs** [comment]
`Running wfl.calculators.generic._run_autopara_wrappable with num_python_subprocesses=10, um_inputs_per_python_subprocess=1` I think it works because in the end ase just calls VASP with a command like `srun -N 1 --ntasks=72 bin/vasp_std`. So all the ase-vasp parcers and waiting proccesses probably run on one node (which should not matter since they are not expensive) and slurm in combination with VASP does the rest of the parallelization (and enables me to use multiple nodes).

So in my case the multiprocessing pool is only sending VASP run commands , waiting and retriving the results in an efficents way.

**Comment by Felixrccs** [comment]
>  The latter is possible, but you may want to check to make sure you're not wasting 9 nodes.

No, I checked that: slurm gives me a 96-99% CPU utilization back and I also did a benchmark to check if I get the wanted time reduction. It also works if I do 2 subprocesses with `srun -N 4` on 8 nodes. If I give the job fewer nodes than I have wfl subprocess slurm complains that it cannot allocate the computational resources.

So Im confident that everything works as intended.

**Comment by bernstei** [comment]
Weird - definitely doesn't seem to work on our slurm system.  If I request an 8 node allocation and run 8 `srun -N 1 --ntasks...` they all run on the same node.  Maybe something about how slurm is configured?
